### PR TITLE
Registration and host_init_config templates previews

### PR DESCRIPTION
**Pull request by stejskalleos**
* Make Global registration template previewable
* Do not require host selection for registration templates
* Show managed & unmanaged hosts for host_init_config template preview

### CODE REVIEW

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 1]
```diff
@@ -171,9 +171,9 @@ if [ -f /etc/redhat-release ]; then
     # Configure subscription-manager
     test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
     subscription-manager config \
-      --server.hostname="<%= @rhsm_url.host %>" \
-      --server.port="<%= @rhsm_url.port %>" \
-      --server.prefix="<%= @rhsm_url.path %>" \
+      --server.hostname="<%= @rhsm_url.host if @rhsm_url %>" \
+      --server.port="<%= @rhsm_url.port if @rhsm_url %>" \
+      --server.prefix="<%= @rhsm_url.path if @rhsm_url %>" \
```
Comment (ezr-ondrej) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:176]: I'd like a safe navigation better here, it is more clear then that the params will be empty is that what we want, for should we omit them altogether `unless @rhsm_url`?
```suggestion
      --server.hostname="<%= @rhsm_url&.host %>" \
      --server.port="<%= @rhsm_url&.port %>" \
      --server.prefix="<%= @rhsm_url%.path %>" \
```

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 2]
```diff
@@ -193,7 +193,7 @@ if [ -f /etc/redhat-release ]; then
     fi
 
     subscription-manager register <%= '--force' if truthy?(@force) %> \
-      --org='<%= @organization.label %>' \
+      --org='<%= @organization.label if @organization %>' \
```
Comment (ezr-ondrej) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:196]: Won't the registration fail without org anyway? If so, we probably want to omit the registration without the org, don't we?
```suggestion
      --org='<%= @organization&.label %>' \
```

File: webpack/assets/javascripts/foreman_template_kind.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable jquery/no-val */
+
+import $ from 'jquery';
+
+import store from './react_app/redux';
+import * as EditorActions from './react_app/components/Editor/EditorActions';
+
+export const initTemplateKind = () => {
+  const value = $('#provisioning_template_template_kind_id').val();
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/foreman_template_kind.js]: We want to get rid of jquery in the future, so if we do not need to, it's better not to use it, these are quite simple without jquery:

```suggestion
  const value = document.getElementById('provisioning_template_template_kind_id').value;
```

File: webpack/assets/javascripts/foreman_template_kind.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  store.dispatch(EditorActions.changeTemplateKind(value));
+};
+
+export const inputOnChange = input => {
+  store.dispatch(EditorActions.changeTemplateKind($(input).val()));
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/foreman_template_kind.js]:
```suggestion
  store.dispatch(EditorActions.changeTemplateKind(input.value));
```

File: app/helpers/templates_helper.rb [hunk 5]
```diff
@@ -103,4 +103,10 @@ def resource_value_options(resource_type)
   def resource_permission(resource_type)
     "view_#{resource_type.demodulize.underscore.pluralize}".to_sym
   end
+
+  def template_class(template)
```
Comment (ezr-ondrej) [inline app/helpers/templates_helper.rb]: Could we be a little bit more specific please? helper method name clash is hard to debug.
Maybe something like `template_class_name` would be at least bit more descriptive... :shrug:

**Review by ezr-ondrej — COMMENTED** [review]
I've few nitpicks, but looks good in general.
I'll test once you implement the ones you agree with :)
I like how quite cleanly you've done the legacy with redux communication :trophy:

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:176]: You can't use safe operator in templates (in safe mode), you'll get error:
```
There was an error rendering the template: wrong number of arguments (given 2, expected 1)

2022-01-19T14:31:00 [I|app|cba3b039] Backtrace for 'Error rendering the  template' error (ArgumentError): wrong number of arguments (given 2, expected 1)
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/safemode-1.3.6/lib/safemode/parser.rb:35:in `process_call'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/ruby2ruby-2.4.4/lib/ruby2ruby.rb:308:in `process_safe_call'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:269:in `block (2 levels) in process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:324:in `error_handler'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:268:in `block in process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:388:in `in_context'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:245:in `process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/ruby2ruby-2.4.4/lib/ruby2ruby.rb:207:in `block in process_block'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/ruby2ruby-2.4.4/lib/ruby2ruby.rb:206:in `map'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/ruby2ruby-2.4.4/lib/ruby2ruby.rb:206:in `process_block'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:269:in `block (2 levels) in process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:324:in `error_handler'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:268:in `block in process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:388:in `in_context'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/sexp_processor-4.16.0/lib/sexp_processor.rb:245:in `process'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/safemode-1.3.6/lib/safemode/parser.rb:10:in `jail'
 cba3b039 | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/safemode-1.3.6/lib/safemode.rb:49:in `eval'
 cba3b039 | /home/vagrant/foreman/app/services/foreman/renderer/safe_mode_renderer.rb:7:in `render'
 cba3b039 | /home/vagrant/foreman/app/services/foreman/renderer/base_renderer.rb:18:in `render'
 cba3b039 | /home/vagrant/foreman/app/models/template.rb:172:in `render'
 cba3b039 | /home/vagrant/foreman/app/controllers/templates_controller.rb:119:in `safe_render'
 cba3b039 | /home/vagrant/foreman/app/controllers/templates_controller.rb:99:in `preview'
```

It's maybe worth looking into it but I'm not sure it's something I want to do as part of this PR

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:196]: > Won't the registration fail without org anyway? 

Yes that's true but when you do the preview `@organization` variable is nil and preview action fails.

**Comment by stejskalleos** [comment]
[test foreman]

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 1 above)
Comment (ezr-ondrej) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:176]: Ugh, didn't know ok then :+1:

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (ezr-ondrej) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb:196]: Right, get it. It's little bit forced to alter the template just for preview, but these are small things so I'm ok :+1:

File: app/views/templates/_form.html.erb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<%= javascript_tag("$(document).on('ContentLoad', tfm.templateInputs.initTypeChanges)"); %>
+<%= javascript_tag("$(document).on('ContentLoad', tfm.templateInputs.initTypeChanges, tfm.templateKind.initTemplateKind)"); %>
```
Comment (ezr-ondrej) [inline app/views/templates/_form.html.erb]: This breaks `ptable` edit as for ptable there is no kind available.
I think this initialization has to be done in `views/provisioning_templates/_custom_tabs.html.erb`

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
Test failures are related, we've broken the Partition tables with this :)

File: app/models/template.rb [hunk 7]
```diff
@@ -206,13 +206,21 @@ def sync_inputs(inputs)
...
     true
   end
 
+  def registration_template?
+    try(:template_kind)&.name == 'registration'
+  end
+
+  def host_init_config_template?
+    try(:template_kind)&.name == 'host_init_config'
+  end
```
Comment (ezr-ondrej) [inline app/models/template.rb]: Can't we return `false` here and override these methods from `ProvisioningTemplate` where we already have the `template_kind`? Also `support_single_host_render?` could be overriden there bit more conveniently :)

File: app/controllers/templates_controller.rb [hunk 8]
```diff
@@ -87,11 +87,18 @@ def preview
     else
       @template = resource_class.new(params[type_name_plural])
```
Comment (ezr-ondrej) [inline app/controllers/templates_controller.rb:88]: I was trying to see if we could send it as `template_kind_id` and feed it here, but I think `type_name_plural` is not working for some time :facepalm:

File: app/controllers/templates_controller.rb [hunk 9]
```diff
@@ -87,11 +87,18 @@ def preview
...
     end
-    base = @template.class.preview_host_collection
-    @host = params[:preview_host_id].present? ? base.find(params[:preview_host_id]) : base.first
-    if @host.nil?
-      render :plain => _('No host could be found for rendering the template'), :status => :not_found
-      return
+
+    template_kind = TemplateKind.find_by(id: params[:kind]) if params[:kind]
```
Comment (ezr-ondrej) [inline app/controllers/templates_controller.rb]: we could name the param with the full `template_kind_id` at least, but that's entirely optional.

**Review by ezr-ondrej — COMMENTED** [review]
Just some ideas for improvements, feel free to discard them if you do not feel like those make sense to you :+1:

It is working now, so these are just readability improvements :)

**Comment by stejskalleos** [comment]
@ezr-ondrej rebased & updated

**Review by ezr-ondrej — APPROVED** [review]
Thanks @stejskalleos ! :+1:
@ares you were saying you have some tests you want to run this through?

**Review by ezr-ondrej — COMMENTED** [review]
Actually @stejskalleos it works, but in preview I still see a host selection for registration templates. But we can hardly do something about it, right?

**Review by ares — APPROVED** [review]
Works great, even the cloning edge case! Thanks @stejskalleos I'm merging this, probably demo worthy :-) Thanks @ezr-ondrej for the code review too!

**Comment by stejskalleos** [comment]
> but in preview I still see a host selection for registration templates.

Yeah, for new template this would require rewriting of the component, and that's something I really didn't want to do :D But edit of existing one and cloning works as expected

>  probably demo worthy :-) 

I don't know it's kinda bug-fix, isn't it?

**Comment by ezr-ondrej** [comment]
> I don't know it's kinda bug-fix, isn't it?

Kinda, but it's a cool to see it working one :) but up to you ;)

**Pull request merged**